### ISSUE TITLE
Add a session on calibration and conformal prediction (`sessions/calibration.ipynb`)

### ISSUE THREAD

**Issue by yoavram** [yoavram/DataSciPy#13]
## Summary

Add one new session notebook to the `DL2026` branch teaching **calibration** and
**conformal prediction**, plus its `solutions/` counterpart and an `index.ipynb` entry.

Nothing in the course currently asks whether a softmax output means anything. Students
finish `sessions/transfer.ipynb` with a CUB-200 classifier at ~80% top-1 whose training
accuracy reached 1.0, and with **top-5 accuracy** as the only set-valued thing they have
seen. This session turns both into questions:

1. When the model says 0.94, is it right 94% of the time? (calibration)
2. Why 5? Why a fixed size? Can a set come with a guarantee? (conformal prediction)

The spine is one question with three answers of increasing honesty: **a raw softmax means
nothing; a calibrated probability is a frequency over a population; a conformal set
carries a finite-sample marginal coverage guarantee.**

The notebook **trains nothing**. It reuses the fine-tuned checkpoint from
`sessions/transfer.ipynb`, runs inference once to produce logits, caches them, and does
everything else in NumPy.

**Prerequisites:** `sessions/transfer.ipynb` (the model, the official CUB split, top-5
accuracy) and `sessions/mle.ipynb` (negative log-likelihood).

**Target class time:** 50–65 minutes.

---

## Before you start

**Do not run the inference cell.** Producing the logits is ~23,000 EfficientNetV2S forward
passes (5,794 images × the clean pass, three noise severities, and the probe). On GPU that
is a few minutes; on CPU it is hours, and it will look like a hang rather than a failure.
Yoav runs that cell on GPU and produces `../data/cub_effnetv2s_logits.npz`. Write the
cell, leave it in the notebook unexecuted, and **stop and ask him for the cache before
attempting to execute the notebook**. Everything downstream of §0 is NumPy and runs on
CPU in seconds.

This is the same division of labour as the other heavy notebooks on this branch: the
expensive cell is run once, by Yoav, before class.

Also ask:

1. **Does `data/cub_effnetv2s_probe.keras` still exist?** Exercise 2 needs the linear
   probe's logits. If it is gone, drop Exercise 2 and say so in the PR.
2. **Has the cache been produced and the notebook executed with outputs** before it is
   committed? The figures are the teaching material; a notebook committed with empty
   outputs is not done.

Placement in `index.ipynb`: add under **Day 3, immediately after Transfer learning**.
Yoav will reorganise days later; do not spend effort on it.

---

## 1. Logits

No new distributed artifact — nothing is added to `download_data.py`, nothing is
committed. The notebook computes its own logits from the existing checkpoint and caches
them locally, following the checkpoint idiom in `CLAUDE.md`: **both paths stay in the
notebook**, the compute-and-save cell (commented out or guarded) and the load cell.

Reminder: **you do not run the compute-and-save cell** — see "Before you start". Write it,
leave it unexecuted, and work against the cache Yoav gives you.

### 1.1 What to compute

Over the official CUB held-out split as `transfer.ipynb` constructs it
(`train_test_split.txt` flag 0, n = 5,794), cache to `../data/cub_effnetv2s_logits.npz`
(gitignored by `*.npz`):

| key | shape | dtype | contents |
|---|---|---|---|
| `labels` | (5794,) | int32 | class index, 0–199 |
| `image_ids` | (5794,) | int32 | CUB image id, for looking up filenames |
| `finetune` | (5794, 200) | float32 | pre-softmax logits, fine-tuned model |
| `probe` | (5794, 200) | float32 | pre-softmax logits, linear probe |
| `finetune_noise08` | (5794, 200) | float32 | fine-tune logits, Gaussian pixel noise σ = 8 |
| `finetune_noise16` | (5794, 200) | float32 | σ = 16 |
| `finetune_noise32` | (5794, 200) | float32 | σ = 32 |

Class names come from `classes.txt` as in `transfer.ipynb`; do not duplicate them into the
archive.

### 1.2 Two things to get right

**Removing the softmax.** The `species` layer has a softmax activation, so `predict`
returns probabilities and temperature scaling would be applied to the wrong quantity.
After loading, set

```python
model.get_layer('species').activation = keras.activations.linear
```

and predict. This changes inference only; the checkpoint on disk is untouched and
`transfer.ipynb` is unaffected. Do not edit `transfer.ipynb`. Make this a visible teaching
moment in the narrative, not a silent line — "the thing we need is upstream of the softmax"
is the first real idea in the notebook.

**Noise goes on the uint8 RGB array before EfficientNet preprocessing**, not after:

```python
Xn = np.clip(X.astype('float32') + rng.normal(0, sigma, X.shape), 0, 255)
```

with a fixed `SEED`. Reuse the image-loading idiom from `transfer.ipynb` verbatim; do not
write a second loader.

---

## 2. Notebook structure

`sessions/calibration.ipynb`, title **"Calibration and conformal prediction"**. Markdown
narrative between every code cell. Sections as `##` headings.

### §0 Setup

House-style header and first code cell (§4). Load the checkpoint, strip the softmax,
compute or load the logits (§1). Then split the 5,794 held-out images **at random** into a
calibration half and a validation half with a fixed `SEED`:

```python
perm = np.random.default_rng(SEED).permutation(n)
calibration, validation = perm[: n // 2], perm[n // 2 :]
```

Narrative must cover: why the split is random rather than stratified (exchangeability is
the assumption everything below rests on), and one honest caveat — `transfer.ipynb`
plotted `val_accuracy` on these same images, so they are not perfectly untouched. With no
early stopping and no model selection on them the leakage is weak, but students should
hear it said rather than find it themselves.

Use **`calibration` / `validation`** terminology throughout. Never write "test".

### §1 The softmax is not a probability

- Accuracy on the validation half vs the mean maximum softmax probability, printed
  together. The gap is the entire motivation for the session.
- Histogram of the maximum probability.
- Display 3–4 images the model got wrong while reporting > 0.95, with predicted and true
  species names.
- Name it: **accuracy/confidence decoupling**. Cross-entropy training can give a good
  decision boundary without giving numbers that may be read as probabilities of being
  correct. A model that memorised 5,994 training images is exactly the regime where this
  bites.

### §2 Reliability diagram and proper scoring rules

- Define calibration: among predictions made with confidence ≈ *p*, a fraction ≈ *p*
  should be correct.
- **Reliability diagram**: 15 equal-width confidence bins; x = mean confidence in bin,
  y = fraction correct; plot y = x. Below the diagonal is overconfidence. Show bin counts
  (bar widths or a twin histogram) so students see that the top bin holds most of the mass.
- Report **ECE** as a single number, with one sentence noting it depends on the binning
  choice. One sentence only — no exercise around it.
- Report **NLL** and **Brier score**. Connect explicitly to `sessions/mle.ipynb`: the NLL
  here *is* the negative log-likelihood maximised during training. Define **proper scoring
  rule** — a loss minimised in expectation by reporting your true probabilities. Accuracy
  is not one: predicting 0.51 or 0.999 for the correct class scores identically, and log
  loss separates them hardest exactly when the model is confidently wrong.

### §3 Temperature scaling

- Definition: `p_k(T) = softmax(z_k / T)`, one scalar for the whole model.
- Fit `log T` by minimising NLL **on the calibration half only**, with `jax.grad` and a
  short explicit gradient-descent loop — a callback to `sessions/jax.ipynb` and
  `sessions/FFN.ipynb`, and a reminder that "fitting" and "deep learning" are not
  synonyms. Print the fitted `T`; expect `T > 1`.
- Show `argmax` invariance: assert accuracy is **identical** before and after, and say why
  — dividing all logits by a positive scalar cannot reorder them.
- Redraw the reliability diagram; print a small before/after table of NLL, Brier, ECE,
  accuracy.
- **Framing to get right — calibration is not free.** What is free is *accuracy*: no
  prediction changes. What it costs is **data**: ~2,900 held-out images that could have
  been training data, and that must be exchangeable with whatever the model meets in
  deployment. Every method in this notebook is paid for in held-out data, and §6 shows
  what happens when the exchangeability half of that price is not really paid. Do not
  write "calibration is free" anywhere.

### §4 From top-5 to prediction sets

- Open by recalling top-5 accuracy from `transfer.ipynb` and asking what justifies the 5.
  A fixed-size set ignores the image and carries no guarantee.
- **Split conformal**, on the calibration half: score `s_i = 1 − p(y_i | x_i)`; threshold
  `q̂` = the `ceil((n+1)(1−α)) / n` empirical quantile of the scores, with α = 0.1. Spell
  out the finite-sample correction rather than calling `np.quantile` at `1−α` — the `+1`
  is why the guarantee is exact rather than asymptotic.
- Prediction set: `C(x) = {k : p(k | x) ≥ 1 − q̂}`.
- Measure **empirical coverage** on the validation half; it should land near 1 − α.
- Report the **set-size distribution**: histogram, mean, fraction of singletons, fraction
  larger than 5, against the constant 5 of top-5.
- Display 2–3 examples with their sets and species names: a singleton for an easy bird, a
  multi-species set for a hard one. With fine-grained classes the sets come out as sets of
  confusable species, and that interpretability is the payoff.
- One short cell: build the sets from the **uncalibrated** probabilities too. Coverage
  holds either way; the set sizes differ. That is what "distribution-free" means — the
  guarantee never required the network's probabilities to be right, only exchangeability.

### §5 What the guarantee does not say

Prose, no new method:

- It is **marginal**: `P(Y ∈ C(X)) ≥ 1 − α` over the joint draw. It does not say every
  image, or every species, is covered at 1 − α. State this plainly; it is the most common
  misreading.
- **Coverage is not correctness.** A set of eight sparrows containing the true species is
  a statement about a procedure, not a species identification.
- It assumes **exchangeability** between calibration and deployment data — an empirical
  assumption, which the next section breaks.
- At most one short paragraph pointing at MC dropout, deep ensembles and Bayesian neural
  networks as the other axis (epistemic uncertainty), with no code.

### §6 When exchangeability fails

Three cells, ≤10 minutes.

- For σ ∈ {0, 8, 16, 32} (logits already cached), compute: accuracy, mean maximum
  probability, ECE under the temperature fitted on **clean** data, conformal coverage under
  the **clean** `q̂`, and mean set size.
- One figure, two panels: coverage vs σ with a dashed 1 − α line; mean set size vs σ.
- Narrative: coverage falls below 1 − α because the calibration scores no longer describe
  the data, and a frozen `q̂` cannot widen the sets to compensate. The clean-fitted
  temperature does not transfer either. The method is not broken — its stated precondition
  is violated, which is a different thing, and field data violates it routinely.

### §7 Recalibrating under shift

Two or three cells. This is the closing move, and it is the practical answer to §6.

- Take a **small labelled sample from the shifted distribution** — 500 images drawn from
  the calibration half at σ = 16 — and refit both the temperature and `q̂` on it.
- Evaluate on the σ = 16 validation half: coverage returns to ≈ 1 − α, and mean set size
  grows. The model did not improve; the sets became honest about it.
- The lesson, stated explicitly: the repair for broken exchangeability is **more
  exchangeable data**, not a cleverer estimator. 500 labelled examples from the deployment
  distribution buy back the guarantee that zero examples cannot.
- One sentence naming weighted and adaptive conformal prediction as the methods for when
  you cannot label anything from the shifted distribution, with a reference. Do not
  implement them.

### §8 Exercises

In-session exercises, with a matching `solutions/calibration.ipynb`.

1. **α sweep.** Plot mean set size against target coverage for α ∈ {0.01, 0.02, 0.05, 0.1,
   0.15, 0.2}. Where is the elbow, and what does it say about the model?
2. **Probe vs fine-tune.** Fit a temperature for the linear probe; compare NLL, ECE and
   mean set size at α = 0.1 against the fine-tune. Two models can differ far more in set
   size than in top-1 — set size is a model comparison that accuracy cannot express.
3. **Coverage is itself random.** Resample the calibration/validation split 200 times on
   the cached logits and histogram the realised coverage. It centres near 1 − α with
   visible spread: the guarantee is over the draw of the calibration set, not a promise
   about the number you happened to get.

### §9 References and Colophon

Guo et al. 2017, *On Calibration of Modern Neural Networks*, ICML; Angelopoulos & Bates
2023, *A Gentle Introduction to Conformal Prediction and Distribution-Free Uncertainty
Quantification*, FnTML; Romano et al. 2020 (adaptive prediction sets) and Tibshirani et
al. 2019 (weighted conformal) as one-line pointers.

---

## 3. Non-goals — do not add these

- **No class-conditional (Mondrian) conformal prediction.** ~29 held-out images per class
  makes per-class quantiles meaningless here. The marginal-vs-conditional point is made in
  prose in §5 and empirically in §6; that is the whole treatment.
- **No adaptive/regularised prediction sets (APS/RAPS)** beyond the one-line mention.
- **No ECE-bashing exercise.** One caveat sentence in §2, nothing more.
- **No MC dropout, deep ensembles or Bayesian neural networks** beyond the paragraph in §5.
- **No training, no new dataset, no new model, no edits to `transfer.ipynb`.**
- **No changes to `download_data.py`** — the logits are a local cache, not a distributed
  artifact.
- **No `torch`, `tensorflow`, `transformers` or `tensorflow_datasets`**, per `CLAUDE.md`.

---

## 4. House style (match `sessions/transfer.ipynb` exactly)

- Opening markdown cell: `![Py4Eng](../logo.png)`, `# Calibration and conformal
  prediction`, `## Yoav Ram`, then an "In this session we will understand:" bullet list.
- First code cell: `%matplotlib inline`, imports, and the Keras/backend print line.
- Fixed `SEED`; `seaborn.set(style='white', context='talk')`.
- Paths relative to the notebook: `../data/...`.
- `# References` then `# Colophon` (CC BY-NC-SA 4.0 + Python logo), copied verbatim from a
  sibling notebook.
- Markdown narrative between every code cell; no runs of bare code.
- Validation terminology, never "test".
- Commit the notebook **with its outputs** — the figures are the teaching material.
- If you add an import, add it to `requirements.txt`.

---

## 5. Acceptance criteria

- [ ] With `../data/cub_effnetv2s_logits.npz` present, `sessions/calibration.ipynb` runs
      top to bottom from a fresh kernel on CPU in under a minute.
- [ ] The inference cell keeps **both paths** — compute-and-save, and load-from-cache —
      per the `CLAUDE.md` checkpoint idiom, and is left unexecuted by the agent.
- [ ] `.venv/bin/jupyter nbconvert --to notebook --execute --stdout
      sessions/calibration.ipynb > /dev/null` passes with no unintended errors **once the
      cache exists**. Do not run this check before then: without the cache it will attempt
      the full inference pass on CPU and appear to hang.
- [ ] `solutions/calibration.ipynb` exists, runs, and matches the in-session exercises.
- [ ] `index.ipynb` links both as `[Calibration and conformal prediction](...) |
      [solution](...)`, under Day 3 after Transfer learning.
- [ ] `download_data.py` and `transfer.ipynb` are unmodified.
- [ ] Measured numbers appear in the narrative (accuracy, mean confidence, fitted `T`,
      coverage, mean set size, recalibrated coverage) rather than placeholders.
- [ ] Empirical coverage on the clean validation half is within ~0.01 of 1 − α. If it is
      not, stop and report rather than adjusting the quantile until it is.
- [ ] Nothing claims calibration is free, or that conformal coverage is a per-image or
      per-class guarantee.

**Comment by yoavram** [comment]
Implemented in branch "calibration"

**Comment by yoavram** [comment]
Done on `DL2026`, merged in a7af384.

- 3890575 — add a session on calibration and conformal prediction
- bc80e12 — address the review

`sessions/calibration.ipynb` and `solutions/calibration.ipynb` are in and linked from `index.ipynb` on Day 3 immediately after Transfer learning. The notebook trains nothing: it strips the softmax off the `species` layer of the existing fine-tuned CUB checkpoint, caches logits to `../data/cub_effnetv2s_logits.npz` (gitignored, not distributed), and does everything downstream in NumPy. Both paths stay in the notebook — the compute-and-save cell and the load cell.

The two open questions resolved as follows: `data/cub_effnetv2s_probe.keras` still exists, so Exercise 2 is in rather than dropped; and the cache was produced on GPU and the notebook committed with its outputs, so every figure in it is real.

The spine survived the measurement: raw softmax, calibrated probability, conformal set with a finite-sample marginal coverage guarantee.
